edn/src/matcher.rs

```diff
@@ -0,0 +1,541 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::HashMap;
+use std::cell::RefCell;
+use itertools::diff_with;
+
+use symbols;
+use types::Value;
+
+/// A trait defining pattern matching rules for any given pattern of type `T`.
+trait PatternMatchingRules<'a, T> {
+    /// Return true if the given pattern matches an arbitrary value.
+    fn matches_any(pattern: &T) -> bool;
+
+    /// Return the placeholder name if the given pattern matches a placeholder.
+    fn matches_placeholder(pattern: &'a T) -> Option<(&'a String)>;
+}
```

Comment:
I was wondering why my initial trait sketch was wrong, but of course -- you take (or compare to) the corresponding `Value` from the _input_.